src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs

```diff
@@ -3,227 +3,372 @@
+                    return 1;
                 }
                 else
                 {
-                    // case 3(a): source is empty array, so can't perform conversion
-                    return null;
+                    return 0;
                 }
             }
+        }
 
-            // case 4: both destination + source type are single elements, so convert
-            return ConvertSimpleType(culture, value, destinationType);
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            var other = obj as ValueProviderResult?;
+            return other.HasValue ? Equals(other.Value) : false;
         }
 
-        private object ConvertSimpleType(CultureInfo culture, object value, Type destinationType)
+        /// <inheritdoc />
+        public bool Equals(ValueProviderResult other)
         {
-            if (value == null || value.GetType().IsAssignableFrom(destinationType))
+            if (Length != other.Length)
             {
-                return value;
+                return false;
             }
+            else
+            {
+                var x = (string[])this;
+                var y = (string[])other;
+                for (var i = 0; i < x.Length; i++)
+                {
+                    if (!string.Equals(x[i], y[i], StringComparison.Ordinal))
+                    {
+                        return false;
+                    }
+                }
+                return true;
+            }
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            return ((string)this)?.GetHashCode() ?? 0;
+        }
+
+        /// <inheritdoc />
+        public override string ToString()
+        {
+            return (string)this;
+        }
+
+        /// <inheritdoc />
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
+
+        /// <inheritdoc />
+        IEnumerator<string> IEnumerable<string>.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
 
-            // In case of a Nullable object, we try again with its underlying type.
-            destinationType = UnwrapNullableType(destinationType);
+        /// <inheritdoc />
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
 
-            // if this is a user-input value but the user didn't type anything, return no value
-            var valueAsString = value as string;
-            if (valueAsString != null && string.IsNullOrWhiteSpace(valueAsString))
+        /// <summary>
+        /// Converts the provided <see cref="ValueProviderResult"/> into a comma-separated string containing all
+        /// submitted values.
+        /// </summary>
+        /// <param name="result">The <see cref="ValueProviderResult"/>.</param>
+        public static implicit operator string(ValueProviderResult result)
```

Comment:
Suggest leaving the casts as explicit. Will make the callers more readable and you seem to already be using `(string)` or `(string[])` where necessary.